### ISSUE TITLE
task model: transitive instance/field inference not working 

### ISSUE THREAD

**Issue by jmesserly** [dart-lang/dev_compiler#347]
The following test used to pass but now does not:
```dart
  testChecker(
      'can infer also from static and instance fields (flag on)',
      {
        '/a.dart': '''
          import 'b.dart';
          class A {
            static final a1 = B.b1;
            final a2 = new B().b2;
          }
      ''',
        '/b.dart': '''
          class B {
            static final b1 = 1;
            final b2 = 1;
          }
      ''',
        '/main.dart': '''
          import "a.dart";

          test1() {
            int x = 0;
            // inference in A now works.
            x = A.a1;
            x = new A().a2;
          }
    '''
      },
      inferTransitively: true);
```
errors:
```
00:01 +15 -1: can infer also from static and instance fields (flag on)
  Unexpected errors reported by checker:
  
  info: [DynamicCast] line 6, column 17 of /main.dart: A.a1 (dynamic) will need runtime check to cast to type int
              x = A.a1;
                  ^^^^
  info: [DynamicCast] line 7, column 17 of /main.dart: new A().a2 (dynamic) will need runtime check to cast to type int
              x = new A().a2;
                  ^^^^^^^^^^
  package:test              fail
  test/testing.dart 181:7   _ExpectedErrorVisitor.validate
  test/testing.dart 123:53  testChecker.<fn>
```

**Comment by jmesserly** [comment]
this also seems to affect "inference in cycles is deterministic"

**Comment by jmesserly** [comment]
also affects "infer if complex expressions read possibly inferred field"

**Comment by leafpetersen** [comment]
The second of the two errors here is #354 .  The first of the two errors I believe is fixed in https://codereview.chromium.org/1370793002 .

**Comment by jmesserly** [comment]
Yup, I can confirm the first is fixed by your CL. Closing this
